src/types/poly_func.rs

```diff
@@ -0,0 +1,613 @@
+)]
+#[display(
+    fmt = "forall {}. {}",
+    "params.iter().map(ToString::to_string).join(\" \")",
+    "body"
+)]
+pub struct PolyFuncType {
+    /// The declared type parameters, i.e., these must be instantiated with
+    /// the same number of [TypeArg]s before the function can be called. Note that within
+    /// the [Self::body], variable (DeBruijn) index 0 is element 0 of this array, i.e. the
+    /// variables are bound from right to left.
+    ///
+    /// [TypeArg]: super::type_param::TypeArg
+    pub(crate) params: Vec<TypeParam>,
+    /// Template for the function. May contain variables up to length of [Self::params]
+    pub(super) body: FunctionType,
+}
+
+impl From<FunctionType> for PolyFuncType {
+    fn from(body: FunctionType) -> Self {
+        Self {
+            params: vec![],
+            body,
+        }
+    }
+}
+
+impl PolyFuncType {
+    /// Create a new PolyFuncType and validates it. (This will only succeed
+    /// for outermost PolyFuncTypes i.e. with no free type-variables.)
+    /// The [ExtensionRegistry] should be the same (or a subset) of that which will later
+    /// be used to validate the Hugr; at this point we only need the types.
+    ///
+    /// #Errors
+    /// Validates that all types in the schema are well-formed and all variables in the body
+    /// are declared with [TypeParam]s that guarantee they will fit.
+    pub fn new_validated(
+        params: impl Into<Vec<TypeParam>>,
+        body: FunctionType,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<Self, SignatureError> {
+        let params = params.into();
+        body.validate(extension_registry, &params)?;
+        Ok(Self { params, body })
+    }
+
+    pub(super) fn validate(
+        &self,
+        reg: &ExtensionRegistry,
+        external_var_decls: &[TypeParam],
+    ) -> Result<(), SignatureError> {
+        let mut v; // Declared here so live until end of scope
+        let all_var_decls = if self.params.is_empty() {
+            external_var_decls
+        } else {
+            // Type vars declared here go at lowest indices (as per DeBruijn)
+            v = self.params.clone();
+            v.extend_from_slice(external_var_decls);
+            v.as_slice()
+        };
+        self.body.validate(reg, all_var_decls)
+    }
+
+    pub(super) fn substitute(&self, t: &impl Substitution) -> Self {
+        if self.params.is_empty() {
+            // Avoid using complex code for simple Monomorphic case
+            return self.body.substitute(t).into();
+        }
+        PolyFuncType {
+            params: self.params.clone(),
+            body: self.body.substitute(&InsideBinders {
+                skip_lowest: self.params.len(),
+                underlying: t,
+            }),
+        }
+    }
+
+    /// (Perhaps-partially) instantiates this [PolyFuncType] into another with fewer binders.
+    /// Note that indices into `args` correspond to the same index within [Self::params],
```

Comment:
The alternative - better in theory but I feel more surprising to users - would be to make the end of the `args` list line up with the end of `params`, consistent with them both being Bwd lists...